osf/models/mixins.py

```diff
@@ -329,9 +329,17 @@ def add_node_link(self, node, auth, save=True):
         # Fail if node already in nodes / pointers. Note: cast node and node
         # to primary keys to test for conflicts with both nodes and pointers
         # contained in `self.nodes`.
+        if (self._id == node._id):
+            raise ValueError(
+                'Cannot link node \'{}\' to itself.'.format(self._id)
+            )
         if NodeRelation.objects.filter(parent=self, child=node, is_node_link=True).exists():
             raise ValueError(
-                'Link to node {0} already exists'.format(node._id)
+                'Target Node \'{}\' already pointed to by \'{}\'.'.format(node._id, self._id)
+            )
+        if NodeRelation.objects.filter(parent=self, child=node, is_node_link=False).exists():
+            raise ValueError(
+                'Target Node \'{}\' is already a child of \'{}\'.'.format(node._id, self._id)
```

Comment:
I'm thinking we shouldn't do two queries on NodeRelation back-to-back to reduce # of queries - maybe just filter on `parent` and `child` - so we see if a NodeRelation exists, and then from there you can check the first result, whether is_node_link is True or False and throw an error from there.